crates/trigger-redis/src/lib.rs

```diff
@@ -1,203 +1,226 @@
+                .entry(address)
+                .or_default()
+                .entry(channel)
+                .or_default()
+                .push(component_id);
+        }
+
+        // Start subscriber(s)
+        let trigger_app = Arc::new(trigger_app);
+        let mut subscriber_tasks = Vec::new();
+        for (address, channel_components) in server_channel_components {
+            let subscriber = Subscriber::new(address, trigger_app.clone(), channel_components)?;
+            let task = tokio::spawn(subscriber.run_listener());
+            subscriber_tasks.push(task);
+        }
 
-        result?
+        // Wait for any task to complete
+        let (res, _, _) = futures::future::select_all(subscriber_tasks).await;
+        res?
     }
 }
 
-impl RedisTrigger {
-    // Handle the message.
-    #[instrument(name = "spin_trigger_redis.handle_message", skip(self, channel_components, msg),
-        err(level = Level::INFO), fields(otel.name = format!("{} receive", msg.get_channel_name()),
-        otel.kind = "consumer", messaging.operation = "receive", messaging.system = "redis"))]
-    async fn handle(
-        &self,
-        address: &str,
-        channel_components: &ChannelComponents,
-        msg: redis::Msg,
-    ) -> Result<()> {
-        let channel = msg.get_channel_name();
-        tracing::info!("Received message on channel {address}:{:?}", channel);
-
-        if let Some(component_ids) = channel_components.get(channel) {
-            let futures = component_ids.iter().map(|id| {
-                tracing::trace!("Executing Redis component {id:?}");
-                SpinRedisExecutor.execute(&self.engine, id, channel, msg.get_payload_bytes())
-            });
-            let results: Vec<_> = join_all(futures).await.into_iter().collect();
-            let errors = results
-                .into_iter()
-                .filter_map(|r| r.err())
-                .collect::<Vec<_>>();
-            if !errors.is_empty() {
-                return Err(anyhow!("{errors:#?}"));
-            }
-        } else {
-            tracing::debug!("No subscription found for {:?}", channel);
-        }
-        Ok(())
-    }
+/// Maps <channel> -> <component IDs>
+type ChannelComponents = HashMap<String, Vec<String>>;
 
-    async fn run_listener(
-        &self,
+/// Subscribes to channels from a single Redis server.
+struct Subscriber {
+    client: Client,
+    trigger_app: Arc<TriggerApp<RedisTrigger>>,
+    channel_components: ChannelComponents,
+}
+
+impl Subscriber {
+    fn new(
         address: String,
+        trigger_app: Arc<TriggerApp<RedisTrigger>>,
         channel_components: ChannelComponents,
-    ) -> Result<()> {
-        tracing::info!("Connecting to Redis server at {}", address);
-        let mut client = Client::open(address.to_string())?;
-        let mut pubsub = client
-            .get_async_connection()
+    ) -> anyhow::Result<Self> {
+        let client = Client::open(address)?;
```

Comment:
Is there a reason to open the client connection here instead of in `run_listener`?

Reply:
`open` is a lie - its just parsing the `address`